contracts/registrar/contract/registrar.sol

```diff
@@ -0,0 +1,272 @@
+                sig:  _sig
+            }));
+        } else {
+            // New checkpoint announcement
+            pending_proposal.count += 1;
+            pending_proposal.index = _sectionIndex;
+            pending_proposal.votemap[_hash].push(Vote({
+                addr: msg.sender,
+                sig:  _sig
+            }));
+        }
+        if (pending_proposal.votemap[_hash].length < threshold) {
+           return true;
+        }
+        checkpoints[_sectionIndex] = _hash;
+        register_height[_sectionIndex] = block.number;
+        latest = _sectionIndex;
+
+        bytes memory sigs;
+        for (uint idx = 0; idx < threshold; idx++) {
+            sigs = abi.encodePacked(sigs, pending_proposal.votemap[_hash][idx].sig);
+        }
+        emit NewCheckpointEvent(_sectionIndex, _hash, sigs);
+        deletePending();
+        return true;
+    }
+
+    /**
+     * @dev Get all admin addresses
+     * @return address list
+     */
+    function GetAllAdmin()
+    public
+    view
+    returns(address[] memory)
+    {
+        address[] memory ret = new address[](adminList.length);
+        for (uint i = 0; i < adminList.length; i++) {
+            ret[i] = adminList[i];
+        }
+        return ret;
+    }
+
+    /**
+     * @dev Get the detail of pending proposal
+     * @return checkpoint index
+     * @return signers who have submitted checkpoint announcement
+     * @return hashes corresponding checkpoint hash
+     */
+    function GetPending()
+    public
+    view
+    returns(uint, address[] memory, bytes32[] memory)
+    {
+        uint idx = 0;
+        address[] memory addr = new address[](pending_proposal.count);
+        bytes32[] memory hashes = new bytes32[](pending_proposal.count);
+        for (uint i = 0; i < adminList.length; i++) {
+            bytes32 h = pending_proposal.usermap[adminList[i]];
+            if (h != "") {
+                addr[idx] = adminList[i];
+                hashes[idx] = h;
+                idx += 1;
+            }
+        }
+        return (pending_proposal.index, addr, hashes);
+    }
+
+    /**
+     * @dev Clear pending proposal
+     */
+    function deletePending()
+    private
+    {
+        for (uint i = 0; i < adminList.length; i++) {
+            bytes32 h = pending_proposal.usermap[adminList[i]];
+            if (h != "") {
+                delete pending_proposal.votemap[h];
+                delete pending_proposal.usermap[adminList[i]];
+            }
+        }
+        delete pending_proposal;
+    }
+
+    /*
+        Fields
+    */
+    // Inflight new stable checkpoint proposal.
+    PendingProposal pending_proposal;
+
+    // A map of admin users who have the permission to update CHT and bloom Trie root
+    mapping(address => bool) admins;
+
+    // A list of admin users so that we can obtain all admin users.
+    address[] adminList;
+
+    // Registered checkpoint information
+    mapping(uint => bytes32) checkpoints;
```

Comment:
So, we've been thinking that there's no real reason (correct us if wrong) to maintain a set of old checkpoints. We only ever care about the latest. That restriction would probably reduce the contract complexity a bit.